doc/development/adding_operators.rst

```diff
@@ -0,0 +1,357 @@
+
+    dev = qml.device("default.qubit", wires=["q1", "q2", "q3"])
+
+    @qml.qnode(dev)
+    def circuit(angle):
+        FlipAndRotate(angle, wire_rot="q1", wire_flip="q1")
+        return qml.expval(qml.PauliZ("q1"))
+
+>>> a = np.array(3.14)
+>>> circuit(a)
+-0.9999987318946099
+
+If all gates used in the decomposition have gradient recipes defined,
+we can even compute gradients of circuits that use the new gate without any extra effort.
+
+>>> qml.grad(circuit)(a)
+-0.0015926529164868282
+
+.. note::
+
+    The example of ``FlipAndRotate`` is simple enough that one could write a function
+
+    .. code-block:: python
+
+        def FlipAndRotate(angle, wire_rot, wire_flip=None, do_flip=False):
+            if do_flip:
+                qml.PauliX(wires=wire_flip)
+            qml.RX(angle, wires=wire_rot)
+
+    and call it in the quantum function *as if it was a gate*.
+    However, classes allow much more functionality, such as defining the adjoint gate above,
+    defining the shape expected for the trainable parameter(s), or specifying gradient rules.
+
+Defining special properties of an operator
+##########################################
+
+Apart from the main ``Operator`` class, operators with special methods or representations
+are implemented as subclasses ``Operation``, ``Observable``, ``Channel``,
+``CVOperation`` and ``CVOperation``.
+
+However, unlike many other frameworks, PennyLane does not use class
+inheritance to define fine-grained properties of operators,
+such as whether it is its own self-inverse, if it is diagonal,
+or whether it can be decomposed into Pauli rotations. This avoids changing the inheritance structure
+every time an application needs to query a new property.
+
+Instead, PennyLane uses "attributes", which are bookkeeping classes that list operators
+which fulfill a specific property.
+
+For example, we can create a new attribute, ``pauli_ops``, like so:
+
+>>> from pennylane.ops.qubits.attributes import Attribute
+>>> pauli_ops = Attribute(["PauliX", "PauliY", "PauliZ"])
+
+We can check either a string or an Operation for inclusion in this set:
+
+>>> qml.PauliX(0) in pauli_ops
+True
+>>> "Hadamard" in pauli_ops
+False
+
+We can also dynamically add operators to the sets at runtime. This is useful
+for adding custom operations to the attributes such as ``composable_rotations``
+and ``self_inverses`` that are used in compilation transforms. For example,
+suppose you have created a new operation ``MyGate``, which you know to be its
+own inverse. Adding it to the set, like so
+
+>>> from pennylane.ops.qubits.attributes import self_inverses
+>>> self_inverses.add("MyGate")
+
+Attributes can also be queried by devices to use special tricks that allow more efficient
+implementations. The onus is on the contributors of new operators to add them to the right attributes.
+
+.. note::
+
+    The attributes for qubit gates are currently found in ``pennylane/ops/qubit/attributes.py``.
```

Comment:
```suggestion
    The attributes for qubit gates are currently found in ``pennylane/ops/qubit/attributes.py``.
    
    Included attributes are listed in the ``Operation``
    `documentation <https://pennylane.readthedocs.io/en/latest/code/qml_operation.html#operation-attributes>`__.
```
(we could also add a sphinx label to the linked section title so that we could use `:ref:`)